mcu/src/mcu/uart_reader.py

```diff
@@ -0,0 +1,455 @@
+            font=("Helvetica", 10),
+        ).grid(row=3, column=0, sticky=tk.W)
+
+        ttk.Label(frame, text="Select COM Port:").grid(row=1, column=0, sticky=tk.W)
+        self.port_combobox = ttk.Combobox(frame, textvariable=self.port_var)
+        self.refresh_ports()
+        self.port_combobox.grid(row=1, column=1, sticky=(tk.W, tk.E))
+
+        if len(self.port_combobox["values"]) > 1:
+            self.port_combobox.current(DEFAULT_PORT_INDEX)
+
+        self.refresh_button = ttk.Button(
+            frame, text="Refresh", command=self.refresh_ports
+        )
+        self.refresh_button.grid(row=1, column=2, sticky=(tk.W, tk.E))
+
+        ttk.Label(frame, text="Baudrate:").grid(row=2, column=0, sticky=tk.W)
+        ttk.Entry(frame, textvariable=self.baudrate_var).grid(
+            row=2, column=1, sticky=(tk.W, tk.E)
+        )
+
+        ttk.Label(frame, text="Frequency Sampling:").grid(row=3, column=0, sticky=tk.W)
+        ttk.Entry(frame, textvariable=self.freq_sampling_var).grid(
+            row=3, column=1, sticky=(tk.W, tk.E)
+        )
+
+        ttk.Label(frame, text="Max ADC Value:").grid(row=4, column=0, sticky=tk.W)
+        ttk.Entry(frame, textvariable=self.val_max_adc_var).grid(
+            row=4, column=1, sticky=(tk.W, tk.E)
+        )
+
+        ttk.Label(frame, text="VDD:").grid(row=5, column=0, sticky=tk.W)
+        ttk.Entry(frame, textvariable=self.vdd_var).grid(
+            row=5, column=1, sticky=(tk.W, tk.E)
+        )
+
+        ttk.Checkbutton(frame, text="Enable Logging", variable=self.log_var).grid(
+            row=6, column=0, columnspan=2, sticky=tk.W
+        )
+        ttk.Checkbutton(
+            frame, text="Auto Delete Audio Files", variable=self.auto_delete_audio_var
+        ).grid(row=6, column=1, columnspan=2, sticky=tk.W)
+
+        self.connect_button = ttk.Button(
+            frame, text="Connect", command=self.toggle_connection
+        )
+        self.connect_button.grid(row=7, column=0, columnspan=2, sticky=(tk.W, tk.E))
+
+        self.status_label = ttk.Label(
+            frame, text="Status: Disconnected", foreground="red"
+        )
+        self.status_label.grid(row=8, column=0, columnspan=2, sticky=(tk.W, tk.E))
+
+        self.console = scrolledtext.ScrolledText(
+            frame, wrap=tk.WORD, width=50, height=10
+        )
+        self.console.grid(row=9, column=0, columnspan=2, sticky=(tk.W, tk.E))
+
+    def print_stuff(self, stuff, type="data"):
+        if type == "error":
+            self.console.insert(tk.END, stuff + "\n", "error")
+            self.console.tag_config("error", foreground="red")
+        elif type == "data":
+            self.console.insert(tk.END, stuff + "\n", "data")
+        self.console.see(tk.END)
+        # Additional print to console and log file
+        print(stuff)
+        if self.log_var.get():
+            with open("uart_reader.log", "a") as f:
+                f.write(stuff + "\n")
```

Comment:
If you use the `logging` module (which we recommend, see my main comment), you can set up a text box that will show the logs (instead of only displaying them in the terminal). See: https://stackoverflow.com/questions/13318742/python-logging-to-tkinter-text-widget